addon/components/gavant-table.ts

```diff
@@ -0,0 +1,532 @@
+     * @type {string} height
+     * @memberof GavantTableComponent
+     */
+    @computed('tableHeight')
+    get height(): SafeString {
+        return htmlSafe(this.tableHeight ? `height: ${this.tableHeight};` : '');
+    }
+
+    /**
+     * CP that returns a class string that includes all of
+     * the enabled properties
+     *
+     * @readonly
+     * @type {string} tableClassNames
+     * @memberof GavantTableComponent
+     */
+    @computed('tableClass', 'showHeader', 'stripedRows', 'hoverableRows', 'clickableRows', 'small')
+    get tableClassNames(): string {
+        const classNames = A([this.tableClass]);
+
+        if (!this.showHeader) {
+            classNames.pushObject('table-no-header');
+        }
+
+        if (this.stripedRows) {
+            classNames.pushObject('table-striped');
+        }
+
+        if (this.hoverableRows) {
+            classNames.pushObject('table-hover');
+        }
+
+        if (this.clickableRows) {
+            classNames.pushObject('table-clickable-rows');
+        }
+
+        if (this.small) {
+            classNames.pushObject('table-sm');
+        }
+
+        return classNames.join(' ');
+    }
+
+    /**
+     * CP that returns a new fillColumnIndex when the current
+     * fillColumnIndex is out of range
+     *
+     * @readonly
+     * @type {(number | null)} adjustedFillColumnIndex
+     * @memberof GavantTableComponent
+     */
+    @computed('fillColumnIndex', 'visibleColumns.length')
+    get adjustedFillColumnIndex(): number | null {
+        if (this.fillColumnIndex) {
+            return this.fillColumnIndex > this.visibleColumns.length - 1 ? 0 : this.fillColumnIndex;
+        } else {
+            return this.fillColumnIndex;
+        }
+    }
+
+    /**
+     * Updates the visible columns when the source columns array changes
+     *
+     * @readonly
+     * @type {*} any
+     * @memberof GavantTableComponent
+     */
+    @observes('columns.[]')
+    onColumnsChange(): any {
+        return scheduleOnce('afterRender', this, 'updateColumnVisibility');
+    }
+
+    //methods
+
+    /**
+     * Initialize the table component's `visibleColumns`
+     *
+     * @memberof GavantTableComponent
+     */
+    init() {
+        super.init();
+        this.columns.forEach((col) => {
+            set(col, 'isVisible', true);
+        });
+    }
+
+    /**
+     * Handles initial column visibility manipulation when the table is first rendered
+     *
+     * @memberof GavantTableComponent
+     */
+    didInsertElement() {
+        super.didInsertElement();
+        if (this.constrainColumnsToFit) {
+            scheduleOnce('afterRender', this, 'updateColumnVisibility');
+        }
+    }
+
+    /**
+     * Handle component container element resize events and hide/show columns as needed
+     *
+     * @memberof GavantTableComponent
+     */
+    debouncedDidResize() {
+        if (this.constrainColumnsToFit) {
+            //reset pan position on container resize
```

Comment:
Should probably move these comments into the main comment section of the method so they are visible